File: mazes-for-programmers/python/sidewinder.py
from random import choice, randint
from grid import Grid


class Sidewinder:
    @staticmethod
    def on(grid):
        for row in grid.each_row():
            run = []
            for cell in row:
                run.append(cell)
                at_eastern_boundary = cell.east is None
                at_northern_boundary = cell.north is None
                should_close_out = at_eastern_boundary or (
                    not at_northern_boundary and randint(0, 2) == 0
                )
                if should_close_out:
                    member = choice(run)
                    if member.north:
                        member.link(member.north)
                    run = []
                else:
                    cell.link(cell.east)
        return grid


if __name__ == "__main__":
    grid = Grid(10, 10)
    Sidewinder().on(grid)
    print(grid)

    grid.to_png()

File: mazes-for-programmers/kotlin/src/main/kotlin/io/github/ocirne/mazes/algorithms/Sidewinder.kt
package io.github.ocirne.mazes.algorithms

import io.github.ocirne.mazes.grids.*
import io.github.ocirne.mazes.grids.CartesianGrid.CartesianCell
import io.github.ocirne.mazes.grids.CartesianGrid.CartesianMaze
import kotlin.random.Random.Default.nextBoolean

class Sidewinder : PassageCarver {

    override fun on(gridProvider: GridProvider, startAt: (Maze) -> Cell): Maze {
        val grid = gridProvider.forPassageCarver()
        grid as CartesianMaze
        for (row in grid.eachRow()) {
            val run = mutableListOf<CartesianCell>()
            for (cell in row) {
                run.add(cell)
                val atEasternBoundary = cell.east == null
                val atNorthernBoundary = cell.north == null

                val shouldCloseOut = atEasternBoundary || (!atNorthernBoundary && nextBoolean())

                if (shouldCloseOut) {
                    val member = run.random()
                    if (member.north != null) {
                        member.link(member.north as CartesianCell)
                    }
                    run.clear()
                } else if (cell.east != null) {
                    cell.link(cell.east as CartesianCell)
                }
            }
        }
        return grid
    }
}
